adr/ADR-8.md

```diff
@@ -293,6 +303,60 @@ all the known services we ask for all subjects below `$KV.SERVICES.nats.>`. A no
 
+#### API Design notes
+
+The API here represents a minimum, languages can add local flavour to the API - for example one can add `PutUint64()` and `GetUint64()`
+in addition to the `Get()` and `Put()` defined here, but it's important that as far as possible all implementations strive to match
+the core API calls - `Get()`, `Put()` and to a lessor extend `Delete()` and `Purge()` and `Entry` - the rest, admin APIs, and supporting
+calls can be adjusted to be a natural fit in the language and design.  This is in line with existing efforts to harmonize 
+`Subscribe()`, `Publish()` and more.
+
+The design is based on extensive prior art in the industry. During development 20+ libraries in Go, Ruby, Python and Java were reviewed 
+for their design choices and this influenced things like `Get()` returning an `Entry`.
+
+Consul Go:
+
+```go
+func (k *KV) Put(p *KVPair, q *WriteOptions) (*WriteMeta, error){}
+func (k *KV) Get(key string, q *QueryOptions) (*KVPair, *QueryMeta, error)
+```
+
+Here KVPair has properties like `Key`, various sequences, flags, timestamps etc and `Value []byte`.
+
+Dynanic language libraries like Python for example return maps, but there too the return value of a default `get()` operation is a hash
+with all metadata and value stored in `Value` or similar.
+
+Various Java libraries from the Consul site were reviewed, there's a mixed bag but most seem to settle on `getXXX() with `getValue()` being
+a common function name  and it returns a fat value object with lots of metadata. [Ecwid/consul-api](https://github.com/Ecwid/consul-api) is a good example.
+
+etcd Go:
+
+```go
+func Put(ctx context.Context, key, val string, opts ...OpOption) (*PutResponse, error) {}
+func Get(ctx context.Context, key string, opts ...OpOption) (*GetResponse, error) {}
+```
+
+Here `GetResponse` is much fatter than we propose as it also supports ranged queries and so multiple values, each being like our Entry.
+
+The official `jetcd` library has a `get()` that returns a future and an equally fat ranged response.
+
+Various Ruby etcd clients were reviewed, same design around `get` with a object returned.
+
+So the overall consensus is that a `Entry` like entity should be returned, hard to say but indeed `Get(string) ([]byte,error)` would seem
+like a good default design, but in the face of massive evidence that this is just not the chosen design I picked returning a `Entry` as 
+the default behavior.  But leaving it open to languages to add additional helpers.
+
+Given that `get()` as a name isn't impossible in Java - see etcd for example - I think in the interest of harmonized client design we should
+use `get()` wherever possible augmented by other get helpers.
+
+Regarding `Put`, these other APIs do not tend to add other functions like `Create()` or `Update()`, they accept put options, like we do.
+Cases where they want to build for example a CAS wrapper around their KV they will write a wrapper with CAS specific function names and more,
```

Comment:
I commented above on create/update. Redis and Memcache do support this. I think Mongo for instance does not.